src/sparse/impl/KokkosSparse_spmv_impl.hpp

```diff
@@ -502,63 +508,606 @@ struct SPMV_MV_Transpose_Functor {
+      x2 += input.x2;
+      x3 += input.x3;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    value_type get(const ordinal_type vecIdx) {
+      switch (vecIdx) {
+      case 0:
+        return x0;
+      case 1:
+        return x1;
+      case 2:
+        return x2;
+      case 3:
+        return x3;
+      }
+
+      return 0;
+    }
+  };
+
+  struct value3 {
+    value_type x0, x1, x2;
+
+    KOKKOS_INLINE_FUNCTION
+    void operator += (const value3 input) {
+      x0 += input.x0;
+      x1 += input.x1;
+      x2 += input.x2;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator += (volatile const value3 input) volatile {
+      x0 += input.x0;
+      x1 += input.x1;
+      x2 += input.x2;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    value_type get(const ordinal_type vecIdx) {
+      switch (vecIdx) {
+      case 0:
+        return x0;
+      case 1:
+        return x1;
+      case 2:
+        return x2;
+      }
+
+      return 0;
+    }
+  };
+
+  struct value2 {
+    value_type x0, x1;
+
+    KOKKOS_INLINE_FUNCTION
+    void operator += (const value2 input) {
+      x0 += input.x0;
+      x1 += input.x1;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator += (volatile const value2 input) volatile {
+      x0 += input.x0;
+      x1 += input.x1;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    value_type get(const ordinal_type vecIdx) {
+      switch (vecIdx) {
+      case 0:
+        return x0;
+      case 1:
+        return x1;
+      }
+
+      return 0;
+    }
+  };
+
+  SPMV_MV_Functor (const value_type alpha_,
+                   const AMatrix m_A_,
+                   const XVector m_x_,
+                   const value_type beta_,
+                   const YVector m_y_,
+                   const ordinal_type rows_per_team_) :
+    alpha (alpha_), m_A (m_A_), m_x (m_x_),
+    beta (beta_), m_y (m_y_),
+    rows_per_team (rows_per_team_),
+    numVecs (m_x_.extent(1)),
+    vecRemainder (m_x_.extent(1) % 16)
+  {
+    static_assert (static_cast<int> (XVector::rank) == 2,
+                   "XVector must be a rank 2 View.");
+    static_assert (static_cast<int> (YVector::rank) == 2,
+                   "YVector must be a rank 2 View.");
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const team_member& dev) const
+  {
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(dev,0,rows_per_team), [&] (const ordinal_type& loop) {
+        const ordinal_type iRow = static_cast<ordinal_type>(dev.league_rank()) * rows_per_team + loop;
+        if (iRow >= m_A.numRows ()) {
+          return;
+        }
+        const KokkosSparse::SparseRowViewConst<AMatrix> row = m_A.rowConst(iRow);
+        const ordinal_type row_length = static_cast<ordinal_type> (row.length);
+        ordinal_type remainder = vecRemainder;
+
+	for(ordinal_type vecOffset = 0; vecOffset < numVecs - 15; vecOffset += 16) {
+	  value16 sum = {};
+	  Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(dev, row_length),
+				  [&] (const ordinal_type& entryIdx, value16& lsum) {
```

Comment:
`ordinal_type` is probably `int`, so it's better to pass by value than by reference.